src/main/java/bio/terra/workspace/db/ResourceDao.java

```diff
@@ -324,6 +324,21 @@ public void createControlledResource(ControlledResource controlledResource)
           "No cloud context found in which to create a controlled resource");
     }
 
+    // scan entire attributes column for existing controlled resource
+    // TODO: use a dedicated URI column to make this more efficient PF-690
```

Comment:
Scanning/parsing all the controlled resource attributes into memory to do an equality check seems like a lot of work to do on every resource creation. Even if we don't want to make a special URI column, we can still write per-resource type SQL query to check the resource/attributes jsonb for equality.

For buckets that's something like
```
SELECT COUNT(1) FROM resource
WHERE attributes->>bucketName = :bucket_name
AND stewardship_type = :controlled
```

Reply:
I thought about that but didn't want that layer of the stack to know about what kind of SQL we're using or to have that step know anything about the different subclasses (like in a switch statement), but it's certainly more efficient. It is hard to defend a full table scan, though.